### ISSUE TITLE
PageRank and Truck Factor as signals would likely make a good criticality score

### ISSUE THREAD

**Issue by HelgeCPH** [ossf/criticality_score#53]
Hej, I saw your project announced on HackerNews the other day and got excited about it. I read the discussions there and those in the issues. I agree with many of the points that were brought up by the community, i.e., that the signal of the current score favor popularity and that the score currently does not really identify _critical_ projects , with _criticality_ in the sense of the Xkcd comic that you linked in your project announcement, i.e., that many other important (foundational) packages depend on a critical project and that the amount of corresponding maintainers is low, potentially a single person.

I think the cool thing about the formula of the criticality score is, that it is not specific on the signals. That is, it seems to me that by identifying the right signals for the right use case(s) one could address many of the concerns from the community.

Since I had the impression from your statements in the discussions, that you are still exploring were to go precisely with the criticality score project, I thought that I try to build a prototype  ([`critical-projects`](https://github.com/HelgeCPH/critical-projects)) that allows to compute the two signals pagerank and truck factor. Pageranks are computed for packages in dependency graphs of package managers and truck factors are computed from Github projects and compared to the values that are computed by your project.

Based on the [results](https://github.com/HelgeCPH/critical-projects/blob/master/data/output/comparison.csv) it seems to me ,that pagerank and truck factor are suitable signals for identifying critical projects, where _criticality_ means many other packages in the ecosystem (transitively) depend on a critical project and that the amount of corresponding maintainers is low, potentially a single person.

So far, I do not have any values to suggest as weights and thresholds for pagerank and truck factor as signals but I think my current results suggest that they are worth investigating further. In the [README](https://github.com/HelgeCPH/critical-projects/blob/master/README.md), I tried to explain a bit more about the results. Please let me know if it does not make sense and then I will try to explain better.

**Comment by inferno-chromium** [comment]
Pagerank is something we want, just wanted to spend more time on getting this right. This will then subsitute dependents_count param that we use now. Right now, first we need something that works for C/C++ as well, since a large portion of security vulnerabilities are confined to memory unsafe languages.

Truck factor needs more discussion. Like it is more clear in terms of funding that yes, we should be supporting projects that don't get much help. But in terms of importance, it is hard to say something should be prioritized if few dev develop it. Needs more thought.

**Comment by jli** [comment]
Just an FYI, Libraries.io has some experimental analyses along these lines:
https://libraries.io/experiments/unseen-infrastructure
https://libraries.io/experiments/digital-infrastructure
https://libraries.io/experiments/bus-factor

(There used to be a link to these "experiments" from the page footer (a few weeks ago), but there aren't anymore, not sure why not.)

I'm not sure of the details of their ranking algorithm, but I'm guessing it's mostly based on their ["SourceRank" metric](https://docs.libraries.io/overview#sourcerank). The libraries.io source code: https://github.com/librariesio/libraries.io

**Comment by nir-krakauer** [comment]
I agree that considering dependency networks, although more complicated than only looking at individual project statistics, is key for identifying some foundational projects that might otherwise be overlooked.

Also drawing inspiration from [xkcd](https://xkcd.com/2347/), I've been discussing for some time with @jli and others how potential contributors can identify the high-impact open-source projects whose maintenance/development would contribute most to the ecosystem.

A metric for package importance could be [developed](https://github.com/nir-krakauer/rankdemo/blob/main/ranking.pdf) by imposing the requirement that the importance of each package be equal to its intrinsic importance (a 'critical-project' score in the terminology of @inferno-chromium) plus the summed importance of the packages that depend on it (directly or indirectly). The linear-algebra algorithm to which such a criterion leads is quite similar to that of PageRank for scoring nodes in hyperlink  networks.

I wrote a sample implementation in GNU Octave called [rankdemo](https://github.com/nir-krakauer/rankdemo/).

**Comment by josephjacks** [comment]
DevRank exists: https://www2.eecs.berkeley.edu/Pubs/TechRpts/2018/EECS-2018-174.pdf

**Comment by nir-krakauer** [comment]
@josephjacks Yes, DevRank is a variant of PageRank that addresses the slightly different question of which contributors within one project are the most valuable, rather than which projects are the most important within a larger software universe.
